File: rgctl-tests/ecommerce-csharp/src/Ecommerce/Services/OrderService.cs
using Ecommerce.Dto;
using Ecommerce.Models;
using Ecommerce.Exceptions;
using Ecommerce.Repositories;

namespace Ecommerce.Services;

public class OrderService
{
    private readonly OrderRepository _orderRepository;
    private readonly CartService _cartService;
    private readonly AuthService _authService;
    private readonly InventoryRepository _inventoryRepository;

    public OrderService(
        OrderRepository orderRepository,
        CartService cartService,
        AuthService authService,
        InventoryRepository inventoryRepository)
    {
        _orderRepository = orderRepository;
        _cartService = cartService;
        _authService = authService;
        _inventoryRepository = inventoryRepository;
    }

    public async Task<List<OrderDto>> FindMyOrdersAsync()
    {
        var user = await _authService.CurrentUserAsync();
        var orders = await _orderRepository.FindByUserIdOrderByCreatedAtDescAsync(user.Id);
        return orders.Select(ToDto).ToList();
    }

    public async Task<OrderDto> FindByIdAsync(long id)
    {
        var order = await GetOrderAsync(id);
        var user = await _authService.CurrentUserAsync();
        if (order.UserId != user.Id && user.Role != "ADMIN")
        {
            throw new ArgumentException("Access denied");
        }

        return ToDto(order);
    }

    public async Task<OrderDto> CheckoutAsync()
    {
        var user = await _authService.CurrentUserAsync();
        var cart = await _cartService.GetUserCartEntityAsync();
        if (cart.Items.Count == 0)
        {
            throw new ArgumentException("Cart is empty");
        }

        var order = new Order
        {
            UserId = user.Id,
            Status = "CONFIRMED"
        };
        var total = 0m;

        foreach (var cartItem in cart.Items)
        {
            var product = cartItem.Product;
            var inventory = await _inventoryRepository.FindByProductIdAsync(product.Id)
                ?? throw new ArgumentException($"Inventory not found for product: {product.Name}");

            if (inventory.Quantity < cartItem.Quantity)
            {
                throw new ArgumentException($"Insufficient stock for product: {product.Name}");
            }

            inventory.Quantity -= cartItem.Quantity;
            await _inventoryRepository.SaveAsync(inventory);

            order.Items.Add(new OrderItem
            {
                ProductId = product.Id,
                Quantity = cartItem.Quantity,
                UnitPrice = product.Price
            });
            total += product.Price * cartItem.Quantity;
        }

        order.TotalAmount = total;
        var saved = await _orderRepository.SaveAsync(order);
        await _cartService.ClearCartAsync();
        return ToDto(saved);
    }

    private async Task<Order> GetOrderAsync(long id)
    {
        return await _orderRepository.FindByIdAsync(id)
            ?? throw new ResourceNotFoundException($"Order not found: {id}");
    }

    private static OrderDto ToDto(Order order)
    {
        var items = order.Items.Select(item => new OrderItemDto(
            item.Id,
            item.ProductId,
            item.Product.Name,
            item.Quantity,
            item.UnitPrice)).ToList();

        return new OrderDto(order.Id, order.Status, order.TotalAmount, order.CreatedAt, items);
    }
}

File: rgctl-tests/ecommerce-cpp/src/services/order_service.cpp
#include "ecommerce/services/order_service.hpp"
#include "ecommerce/services/cart_service.hpp"
#include "ecommerce/services/product_service.hpp"
#include "ecommerce/services/inventory_service.hpp"
#include "ecommerce/repositories/order_repository.hpp"
#include "ecommerce/types.hpp"

namespace ecommerce::services {

int checkout(sqlite3* db, int user_id) {
    if (!db || user_id <= 0) return -1;
    CartItem items[32];
    int count = 0;
    if (get_user_cart(db, user_id, items, sizeof(items), &count) != 0) return -1;
    if (count == 0) return -2;
    double total = 0;
    for (int i = 0; i < count; i++) {
        Product product{};
        if (get_product(db, items[i].product_id, &product) != 0) continue;
        Inventory inv{};
        if (get_by_product(db, items[i].product_id, &inv) != 0) return -3;
        if (inv.quantity < items[i].quantity) return -4;
        total += product.price * items[i].quantity;
    }
    Order order{};
    order.user_id = user_id;
    order.total = total;
    if (repositories::order_create(db, &order) != 0) return -5;
    clear_cart(db, user_id);
    return order.id;
}

int get_order(sqlite3* db, int id, void* out) {
    return repositories::order_find_by_id(db, id, out);
}

}  // namespace ecommerce::services
